tutorials/src/bin/basic-tutorial-8.rs

```diff
@@ -0,0 +1,260 @@
+    if let Err(err) = gst::init() {
+        eprintln!("Failed to initialize Gst: {}", err);
+        return;
+    }
+
+    let app_source = gst::ElementFactory::make("appsrc", "audio_source").unwrap();
+    let tee = gst::ElementFactory::make("tee", "tee").unwrap();
+    let audio_queue = gst::ElementFactory::make("queue", "audio_queue").unwrap();
+    let audio_convert1 = gst::ElementFactory::make("audioconvert", "audio_convert1").unwrap();
+    let audio_resample = gst::ElementFactory::make("audioresample", "audio_resample").unwrap();
+    let audio_sink = gst::ElementFactory::make("autoaudiosink", "audio_sink").unwrap();
+    let video_queue = gst::ElementFactory::make("queue", "video_queue").unwrap();
+    let audio_convert2 = gst::ElementFactory::make("audioconvert", "audio_convert2").unwrap();
+    let visual = gst::ElementFactory::make("wavescope", "visual").unwrap();
+    let video_convert = gst::ElementFactory::make("videoconvert", "video_convert").unwrap();
+    let video_sink = gst::ElementFactory::make("autovideosink", "video_sink").unwrap();
+    let app_queue = gst::ElementFactory::make("queue", "app_queue").unwrap();
+    let app_sink = gst::ElementFactory::make("appsink", "app_sink").unwrap();
+
+    let pipeline = gst::Pipeline::new("test-pipeline");
+
+    visual.set_property_from_str("shader", "none");
+    visual.set_property_from_str("style", "lines");
+
+    // configure appsrc
+    let info = AudioInfo::new(gst_audio::AudioFormat::S16le, SAMPLE_RATE, 1)
+        .build()
+        .unwrap();
+    let audio_caps = info.to_caps().unwrap();
+    app_source.set_property("caps", &audio_caps).unwrap();
+    app_source.set_property_from_str("format", "time");
+
+    let appsrc = app_source.clone().dynamic_cast::<AppSrc>().expect(
+        "Source element is expected to be an appsrc!",
+    );
+
+    let appsink = app_sink.clone().dynamic_cast::<AppSink>().expect(
+        "Sink element is expected to be an appsink!",
+    );
+
+    let data: Arc<Mutex<CustomData>> =
+        Arc::new(Mutex::new(CustomData::new(appsrc.clone(), appsink.clone())));
+
+    let data_clone = data.clone();
+    appsrc.connect_need_data(move |_, _size| {
+        let data = &data_clone;
+        let mut d = data_clone.lock().unwrap();
+
+        if d.source_id.is_none() {
+            println!("start feeding");
+
+            let data_clone = data.clone();
+            d.source_id = Some(idle_add(move || {
+                let data = &data_clone;
+                let mut samples: Vec<u8> = vec![0; CHUNK_SIZE];
+                let num_samples = CHUNK_SIZE / 2; /* Each sample is 16 bits */
+                let appsrc;
+                let pts;
+
+                {
+                    let mut data = data.lock().unwrap();
+                    let samples = samples.as_mut_slice_of::<i16>().unwrap();
+                    pts = data.num_samples
+                        .mul_div_floor(gst::SECOND, SAMPLE_RATE as u64)
+                        .unwrap();
```

Comment:
If you wait long enough, this will panic as u64 overflows ;)

Reply:
What would be the proper way to handle that?

Reply:
Use `expect` instead of `unwrap` for now, or if it returns `None` (i.e. overflows), you could just go EOS and stop doing anything